activesupport/lib/active_support/cache/memory_store.rb

```diff
@@ -114,45 +131,45 @@ def synchronize(&block) # :nodoc:
       private
         PER_ENTRY_OVERHEAD = 240
 
-        def cached_size(key, entry)
-          key.to_s.bytesize + entry.size + PER_ENTRY_OVERHEAD
+        def cached_size(key, payload)
+          key.to_s.bytesize + payload.bytesize + PER_ENTRY_OVERHEAD
         end
 
         def read_entry(key, **options)
           entry = nil
           synchronize do
-            entry = @data.delete(key)
-            if entry
-              @data[key] = entry
+            payload = @data.delete(key)
+            if payload
+              @data[key] = payload
+              entry = deserialize_entry(payload)
               entry = entry.dup
             end
           end
-          entry&.dup_value!
           entry
         end
 
         def write_entry(key, entry, **options)
-          entry.dup_value!
+          payload = serialize_entry(entry)
           synchronize do
-            return false if options[:unless_exist] && @data.key?(key)
+            return false if @data.key?(key) && options[:unless_exist]
```

Comment:
If I understood you correctly in https://github.com/rails/rails/pull/39770#discussion_r479963921:

```suggestion
            return false if options[:unless_exist] && @data.key?(key)
```